tests/integration/ha/test_failover.py

```diff
@@ -0,0 +1,416 @@
+
+    # ensure data is written in the cluster
+    assert_continuous_writes_increasing(endpoints=endpoints, user=INTERNAL_USER, password=password)
+    stop_continuous_writes()
+    # By default, etcd uses a 1s election timeout before attempting to replace a lost leader
+    # that's why we will miss writes here, and therefore ignore the revision of the key
+    assert_continuous_writes_consistent(
+        endpoints=endpoints, user=INTERNAL_USER, password=password, ignore_revision=True
+    )
+
+
+@pytest.mark.runner(["self-hosted", "linux", "X64", "jammy", "large"])
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_full_cluster_restart(ops_test: OpsTest) -> None:
+    """Make sure the cluster can self-heal after all members went down."""
+    app = (await existing_app(ops_test)) or APP_NAME
+
+    # make sure we have at least two units so we can kill one of them
+    if len(ops_test.model.applications[app].units) < 2:
+        await ops_test.model.applications[app].add_unit(count=1)
+        await wait_until(
+            ops_test,
+            apps=[app],
+            apps_statuses=["active"],
+            units_statuses=["active"],
+            wait_for_exact_units=2,
+        )
+
+    init_units_count = len(ops_test.model.applications[app].units)
+    endpoints = get_cluster_endpoints(ops_test, app)
+    secret = await get_secret_by_label(ops_test, label=f"{PEER_RELATION}.{app}.app")
+    password = secret.get(f"{INTERNAL_USER}-password")
+
+    # start writing data to the cluster
+    start_continuous_writes(endpoints=endpoints, user=INTERNAL_USER, password=password)
+    time.sleep(10)
+
+    # update the restart delay for all units
+    for unit in ops_test.model.applications[app].units:
+        await patch_restart_delay(ops_test, unit_name=unit.name, delay=RESTART_DELAY_PATCHED)
+
+    # axe away the etcd process on all units
+    for unit in ops_test.model.applications[app].units:
+        send_process_control_signal(
+            unit_name=unit.name, model_full_name=ops_test.model_full_name, signal="SIGTERM"
+        )
+
+    # ensure the all cluster members are down
+    for unit in ops_test.model.applications[app].units:
+        unit_endpoint = get_unit_endpoint(ops_test, unit_name=unit.name, app_name=app)
+        assert not is_endpoint_up(unit_endpoint, user=INTERNAL_USER, password=password)
+    logger.info("Cluster is not available after being stopped.")
+
+    logger.info(f"Waiting {RESTART_DELAY_PATCHED}s for service restarts.")
+    time.sleep(RESTART_DELAY_PATCHED)
+
+    # now check the availability and formation of the cluster
+    cluster_members = get_cluster_members(endpoints)
+    assert len(cluster_members) == init_units_count, (
+        f"expected {init_units_count} cluster members, got {len(cluster_members)}"
+    )
+    logger.info(f"Cluster has come back, all {len(cluster_members)} members joined the cluster.")
+
```

Comment:
We should reset the restart delay to the default value on the systemd units in this and the following test